### ISSUE TITLE
multiple Elastix instances/runs ignore `RandomSeed` in ParameterFile

### ISSUE THREAD

**Issue by romangrothausmann** [SuperElastix/elastix#25]
As instructed (https://github.com/SuperElastix/SimpleElastix/issues/122#issuecomment-324314648)  here the port of the issue from SimpleElastix  (https://github.com/SuperElastix/SimpleElastix/issues/122) which seems to be an Elastix issue:

It seems (Simple)Elastix ignores `RandomSeed` set in ParameterList if run multiple times.
Even though I set `RandomSeed` in the ParameterFile, and initialize each Elastix instance with the very same ParameterFile (pM), Elastix seems not to (re-) set its random seed accordingly.  See e.g. these test results from selx: https://github.com/romangrothausmann/elastix_scripts/commit/d2d6d04cfa9aaf6121f35d01e601b8cb65284b96?diff=split#diff-2e0a2595b81a2453a02093ffede7ec7dL1

The first metric value (107.81864) stays the same but the second run of selx yields 99.225277 even though both run with "NoInitialTransform" and the same ParameterFile. It also changes the metric values for all following runs. 

Note that the metric value for the second image pair (L2) changes back to its original value (60.185559) if the first image pair only uses one selx run: https://github.com/romangrothausmann/elastix_scripts/commit/d06bd99f8066c5d1429149412895d3046e35d0a9#diff-2e0a2595b81a2453a02093ffede7ec7dR2

So it seems to me that selx initializes its RNG with a fixed random seed (121212 by default or that given by `RandomSeed` in ParameterFile) only once but not for every new instance/run even if the ParameterFile stays the same (and one would expect the RandomSeed to be reset again).

**Comment by kaspermarstal** [comment]
@mstaring, @stefanklein?

**Comment by romangrothausmann** [comment]
Any ideas how to solve this problem? This non-determinism still seems to persist in selx.

**Comment by mstaring** [comment]
The following code seems relevant:

https://github.com/SuperElastix/elastix/blob/develop/Core/Kernel/elxElastixBase.cxx#L237-L246

@romangrothausmann can you check if just plain elastix also behaves like this? Because I would expect it to work correctly.

@kaspermarstal is the function ElastixBase::BeforeAllBase( void ) also called after each start of the registration, if the registration is started via the library interface?

**Comment by romangrothausmann** [comment]
Thaks @mstaring for looking into this.
> can you check if just plain elastix also behaves like this? Because I would expect it to work correctly.

Hm, I'm not sure how to do that. To my knowledge I can run elastix only for aligning one image pair. So for the following pair I would have to run elastix again, which would initialize the library again, so the problem would not occur. Therefore, I would expect that the problem could only be reproduced outside selx by writing a program that uses the elastix C++ library, which I have not used before but instead went for selx.

**Comment by mstaring** [comment]
I am guessing that  ElastixBase::BeforeAllBase( void ) is not called when using elastix via the library interface, so that the seed is never set or reset.

**Comment by mstaring** [comment]
Hi @romangrothausmann , @kaspermarstal is now busy with some stuff related to the continuous registration challenge, so this random seed issue has quite low priority. 
@kaspermarstal: When you have time later in life, my guess is that the library interface should include calls to  ElastixBase::BeforeAllBase( void ) to set/reset the seed.

**Comment by romangrothausmann** [comment]
Thanks for the info. No problem, I can stick to the grid sampler for reproducible continuation after manual interaction for now. The random point sampler seems to need less manual interactions in my series registration approach though.

**Comment by mstaring** [comment]
I am closing the issue due to inactivity. Feel free to reopen when needed.
